docs/user/kafka/streams.rst

```diff
@@ -0,0 +1,119 @@
+
+  object MessageAction extends Enumeration {
+    type MessageAction = Value
+    val Upsert, Delete = Value
+  }
+
+Read Example
+------------
+
+The following is adapted from the word-count `example`__ in Kafka Streams:
+
+__ https://kafka.apache.org/31/documentation/streams/developer-guide/dsl-api.html#scala-dsl-sample-usage
+
+.. code-block:: scala
+
+    import org.apache.kafka.streams.scala.ImplicitConversions._
+    import org.apache.kafka.streams.scala.serialization.Serdes._
+    import org.locationtech.geomesa.kafka.streams.GeoMesaStreamsBuilder
+
+    // these are the parameters normally passed to DataStoreFinder
+    val params = Map[String, String](
+      "kafka.brokers"    -> "localhost:9092",
+      "kafka.zookeepers" -> "localhost:2181"
+    )
+
+    val builder = GeoMesaStreamsBuilder(params)
+
+    // read in the feature type and turn the attributes into a space-separated string
+    val textLines: KStream[String, String] =
+      builder.stream("my-feature-type")
+          .map((k, v) => (k, v.attributes.map(_.toString.replaceAll(" ", "_")).mkString(" ")))
+
+    val wordCounts: KTable[String, Long] =
+      textLines
+          .flatMapValues(textLine => textLine.split(" +"))
+          .groupBy((_, word) => word)
+          .count()(Materialized.as("counts-store"))
+
+    wordCounts.toStream.to("word-count")
+
+    val topology = builder.build()
+    // construct the streams app as normal
+
+Write Example
+-------------
+
+The following shows how to persist data back to a GeoMesa topic:
+
+.. code-block:: scala
+
+    import org.apache.kafka.streams.scala.ImplicitConversions._
+    import org.apache.kafka.streams.scala.serialization.Serdes._
+    import org.locationtech.geomesa.kafka.streams.GeoMesaMessage
+    import org.locationtech.geomesa.kafka.streams.GeoMesaStreamsBuilder
+
+    // these are the parameters normally passed to DataStoreFinder
+    val params = Map[String, String](
+      "kafka.brokers"    -> "localhost:9092",
+      "kafka.zookeepers" -> "localhost:2181"
+    )
+
+    val builder = GeoMesaStreamsBuilder(params)
+
+    // use the wrapped native streams builder to create an input based on csv records
+    val input: KStream[String, String] =
+      builder.builder.stream[String, String]("input-csv-topic")
+
+    // the columns in the csv need to map to the attributes in the feature type
+    val output: KStream[String, GeoMesaMessage] =
+      input.mapValues(lines => GeoMesaMessage.upsert(lines.split(",")))
+
+    // write the output to GeoMesa - the feature type must already exist
+    builder.to("my-feature-type", output)
+
+    val topology = builder.build()
+    // construct the streams app as normal
+
+Joins and Topic Partitioning
+----------------------------
+
+.. warning::
+
+    Kafka streams operations that require co-partitioning will usually require a re-partition step when used
+    with data from GeoMesa 3.4.x or earlier.
```

Comment:
*Blinks*  Why?

Reply:
we had an old custom partitioner, needed b/c originally we used a timestamp for the kafka message key and we needed to ensure that features with the same FID went to the same topic. We switched to using the FID as the message key a while ago, but we never removed the custom partitioner. with co-partitioning and streams, you have to ensure that the same key in both streams is in the same partition. Since we used a custom partitioner, that won't be the case, so you have to re-partition the stream.